src/mlpack/methods/pca/decomposition_policies/stochastic_method.hpp

```diff
@@ -0,0 +1,299 @@
+
+      // Take ownership of the new data.
+      data = std::move(newData);
+    }
+
+    //! Return the number of functions.
+    size_t NumFunctions() const { return data.n_cols; }
+
+    /*
+     * Evaluate a function for a particular batch-size.
+     *
+     * @param coordinates The function coordinates.
+     * @param begin The first function.
+     * @param batchSize Number of points to process.
+     */
+    double Evaluate(const arma::mat& /* coordinates */,
+                    const size_t /* begin */,
+                    const size_t /* batchSize */)
+    {
+      return pl;
+    }
+
+    /*
+     * Evaluate the gradient of a function for a particular batch-size.
+     *
+     * @param coordinates The function coordinates.
+     * @param begin The first function.
+     * @param gradient The function gradient.
+     * @param batchSize Number of points to process.
+     */
+    void Gradient(const arma::mat& coordinates,
+                  const size_t begin,
+                  arma::mat& gradient,
+                  const size_t batchSize)
+    {
+      gradient = data.cols(begin, begin + batchSize - 1) *
+          data.cols(begin, begin + batchSize - 1).t() * coordinates;
+    }
+
+   private:
+    //! Data matrix..
+    arma::mat data;
+
+    //! Locally-stored pseudo loss.
+    double pl;
+
+    //! For shuffling.
+    arma::Row<size_t> visitationOrder;
+  };
+};
+
+/**
+ * Implementation of the PCA Update policy which wraps the existing optimizer
+ * update policies.
+ */
+template<typename UpdatePolicyType = optimization::VanillaUpdate>
+class PCAUpdate
+{
+ public:
+  /**
+   * Construct the wrapper update policy with the given update policy.
+   *
+   * @param beta1 Exponential decay rate for the first moment estimates.
+   * @param beta2 Exponential decay rate for the weighted infinity norm
+   *        estimates.
+   * @param epsilon Value used to initialise the mean squared gradient
+   *        parameter.
+   */
+  PCAUpdate(const UpdatePolicyType& updatePolicy = UpdatePolicyType()) :
+      updatePolicy(updatePolicy)
+  {
+    /* Nothing to do here. */
+  }
+
+  /**
+   * Adam specific constructor.
+   *
+   * @param beta1 Exponential decay rate for the first moment estimates.
+   * @param beta2 Exponential decay rate for the weighted infinity norm
+   *        estimates.
+   * @param epsilon Value used to initialise the mean squared gradient
+   *        parameter.
+   */
+  PCAUpdate(const double epsilon, const double beta1, const double beta2) :
+      updatePolicy(UpdatePolicyType(epsilon, beta1, beta2))
+  {
+    /* Nothing to do here. */
+  }
+
+  void Initialize(const size_t rows, const size_t cols)
+  {
+    updatePolicy.Initialize(rows, cols);
+  }
+
+  /**
+   * General update step, all parameters are forwarded to the actual
+   * update policy.
+   *
+   * @param iterate Parameters that minimize the function.
+   */
+  template<typename... Targs>
+  void Update(arma::mat& iterate, Targs... Fargs)
```

Comment:
I guess if you want to be really picky it should be `fargs` (or `fArgs`?) not `Fargs`, but I dunno if you want to handle template pack parameters differently. :)